### PR TITLE
Add `failover.recoverByUID` field, support tikv/tiflash/dm-worker recover by this uid

### LINKED ISSUE #3335: `recoverFailover` may be forgotten to unset
## Feature Request

**Is your feature request related to a problem? Please describe:**

#3189 introduces `recoverFailover` field in the spec of TidbCluster to recover from the failover Pods, however, after setting it to `true` and the Pods created due to failover are removed, users have to set it to `false` manually, which is not elegant and users may forget to do that.
We want a more reliable and automatical way to do the recover.

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by better0332**
### What problem does this PR solve?

Add `failover.recoverByUID` field, support tikv/tiflash/dm-worker recover by this uid

Closes #3335

### What is changed and how does it work?

After each round of failover starts, generate a new UID , and then set `status.failoverUID`. if `status.failoverUID==failover.recoverByUID`, this round of failover recovers and then operator clean `status.failoverUID`

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support recovering auto-failover by UID for TiKV, TiFlash, and DM worker
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -2333,3 +2348,11 @@ type TopologySpreadConstraint struct {
 	// See pkg/apis/pingcap/v1alpha1/tidbcluster_component.go#TopologySpreadConstraints()
 	TopologyKey string `json:"topologyKey"`
 }
+
+// Failover contains the failover specification.
+// +k8s:openapi-gen=true
+type Failover struct {
+	// RecoverByUID indicates that Operator can recover by this uid , it takes effect only when set `spec.recoverFailover=false`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:2355]:
```suggestion
	// RecoverByUID indicates that TiDB Operator will recover the failover by this UID,
	// it takes effect only when set `spec.recoverFailover=false`
```

File: pkg/manager/member/dm_worker_failover_test.go [hunk 2]
```diff
@@ -119,33 +128,34 @@ func TestWorkerFailoverFailover(t *testing.T) {
 			expectFn: func(t *testing.T, dc *v1alpha1.DMCluster) {
 				g := NewGomegaWithT(t)
 				g.Expect(len(dc.Status.Worker.FailureMembers)).To(Equal(1))
+				g.Expect(dc.Status.Worker.FailoverUID).To(BeEmpty())
```
Comment (DanielZhangQD) [inline pkg/manager/member/dm_worker_failover_test.go:131]: Why there is a failure member but the uid is empty?

File: pkg/manager/member/dm_worker_failover_test.go [hunk 3]
```diff
@@ -227,37 +239,39 @@ func TestWorkerFailoverFailover(t *testing.T) {
 			expectFn: func(t *testing.T, dc *v1alpha1.DMCluster) {
 				g := NewGomegaWithT(t)
 				g.Expect(len(dc.Status.Worker.FailureMembers)).To(Equal(3))
+				g.Expect(dc.Status.Worker.FailoverUID).To(BeEmpty())
```
Comment (DanielZhangQD) [inline pkg/manager/member/dm_worker_failover_test.go:242]: ditto

File: pkg/manager/member/tikv_failover_test.go [hunk 4]
```diff
@@ -120,6 +125,7 @@ func TestTiKVFailoverFailover(t *testing.T) {
 			expectFn: func(t *testing.T, tc *v1alpha1.TidbCluster) {
 				g := NewGomegaWithT(t)
 				g.Expect(len(tc.Status.TiKV.FailureStores)).To(Equal(1))
+				g.Expect(tc.Status.TiKV.FailoverUID).To(BeEmpty())
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_failover_test.go:128]: Why it's empty with 1 failure store?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 5]
```diff
@@ -1067,6 +1067,100 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			ginkgo.By("Waiting for delete scale-tikv-3 pod because of recover")
+			delPodName := controller.TiKVMemberName(clusterName) + "-3"
+			err = wait.PollImmediate(time.Second*10, 5*time.Minute, func() (bool, error) {
+				_, err := c.CoreV1().Pods(ns).Get(context.TODO(), delPodName, metav1.GetOptions{})
+				if err != nil && apierrors.IsNotFound(err) {
+					return true, nil
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err, "failed to wait for delete scale-tikv-3 pod because of recover")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1162]: Need to failover again and make sure the new failed Pod cannot be recovered automatically.

File: pkg/manager/member/tikv_failover_test.go (hunk 4 above)
Comment (better0332) [reply pkg/manager/member/tikv_failover_test.go:128]: The UID be generated together with append FailureStores

File: pkg/manager/member/dm_worker_failover_test.go (hunk 2 above)
Comment (better0332) [reply pkg/manager/member/dm_worker_failover_test.go:131]: The UID be generated together with append FailureStores, `dm-worker-1` already exist there

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 5 above)
Comment (better0332) [reply tests/e2e/tidbcluster/tidbcluster.go:1162]: done

File: pkg/manager/member/dm_worker_failover_test.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/manager/member/dm_worker_failover_test.go:131]: In this case, if failover occurred before upgrading to the TiDB Operator including this fix, the user still cannot set the recoverByUID, right? What about we also generate the uid for this case, e.g. `if worker.Stage == v1alpha1.DMWorkerStateOffline && time.Now().After(deadline), check and set the status`.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -2333,3 +2348,12 @@ type TopologySpreadConstraint struct {
 	// See pkg/apis/pingcap/v1alpha1/tidbcluster_component.go#TopologySpreadConstraints()
 	TopologyKey string `json:"topologyKey"`
 }
+
+// Failover contains the failover specification.
+// +k8s:openapi-gen=true
+type Failover struct {
+	// RecoverByUID indicates that TiDB Operator will recover the failover by this UID,
+	// it takes effect only when set `spec.recoverFailover=false`
```
Comment (KanShiori) [inline pkg/apis/pingcap/v1alpha1/types.go:2356]: set `spec.recoverFailover=true` ?

Comment (better0332) [reply pkg/apis/pingcap/v1alpha1/types.go:2356]: if `spec.recoverFailover=true`, Operator recover whatever RecoverByUID set

Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/types.go:2356]: I misunderstood 🤣

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 7]
```diff
@@ -1067,6 +1067,211 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 			})
 			framework.ExpectNoError(err, "not clear TiDB failureMembers when scale TiDB to zero")
 		})
+
+		ginkgo.It("should failover tikv and recover by recoverByUID", func() {
+			clusterName := "scale"
```
Comment (KanShiori) [inline tests/e2e/tidbcluster/tidbcluster.go:1072]:
```suggestion
			clusterName := "recover-by-uid"
```

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 8]
```diff
@@ -1067,6 +1067,211 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				_, err := c.CoreV1().Pods(ns).Get(context.TODO(), delPodName, metav1.GetOptions{})
+				if err != nil && apierrors.IsNotFound(err) {
+					return true, nil
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err, "failed to wait for delete scale-tikv-3 pod because of recover")
+		})
+
+		ginkgo.It("should failover tikv and not recover by wrong recoverByUID, recover by recoverFailover eventually", func() {
```
Comment (KanShiori) [inline tests/e2e/tidbcluster/tidbcluster.go:1165]: Do we need an e2e  case to cover this case?
Maybe use a ut to cover it

**Review by KanShiori — APPROVED** [review]

**Pull request merged**